base/compiler/ssair/passes.jl

```diff
@@ -135,7 +137,7 @@ function simple_walk(compact::IncrementalCompact, @nospecialize(defssa#=::AnySSA
 end
 
 function simple_walk_constraint(compact, defidx, typeconstraint = types(compact)[defidx])
-    callback = (pi, _)->isa(pi, PiNode) && (typeconstraint = typeintersect(typeconstraint, pi.typ))
+    callback = (pi, _)->(isa(pi, PiNode) && (typeconstraint = typeintersect(typeconstraint, pi.typ)); false)
```

Comment:
if you're using `isa` checks, should have `@nospecialize` on all copies of this callback definition